pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,296 @@
+package distributor
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const (
+	longPollDuration = 10 * time.Second
+)
+
+// ErrDontOverwrite is used when we want to accept a sample but not change the values in the KVStore when CAS'ing.
+type ErrDontOverwrite struct {
+}
+
+func (e *ErrDontOverwrite) Error() string {
+	return fmt.Sprintf("don't overwrite")
+}
+
+// Track the most recent timestamp for the replica we're
+// accepting samples from in a Prometheus HA setup.
+type replicaTracker struct {
```

Comment:
Can we get rid of this and just use `ReplicaDescription` from the generated protos?